telemetry/app.go

```diff
@@ -0,0 +1,135 @@
+)
+
+// InitAppMetrics initializes the block specific metrics for the application.
+func InitAppMetrics(homeDir string) func(*baseapp.BaseApp) {
+	return func(app *baseapp.BaseApp) {
+		app.SetPrepareProposal(initPrepareProposalHandlerWithMetrics(app))
+		app.SetStreamingManager(initStreamingManagerWithMetrics(app, path.Join(homeDir, "data")))
+	}
+}
+
+// initPrepareProposalHandlerWithMetrics initializes the prepare proposal handler
+// with the app metrics.
+// It gathers the block txs size to emit them as a gauge metric.
+func initPrepareProposalHandlerWithMetrics(app *baseapp.BaseApp) sdk.PrepareProposalHandler {
+	// Create a NoOpMempool for the application and get the default prepare proposal
+	// handler as per NewBaseApp implementation.
+	// See https://github.com/cosmos/cosmos-sdk/blob/v0.50.4/baseapp/baseapp.go#L221
+	app.SetMempool(mempool.NoOpMempool{})
+	abciProposalHandler := baseapp.NewDefaultProposalHandler(app.Mempool(), app)
+	prepareProposalHandler := abciProposalHandler.PrepareProposalHandler()
+
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		// Iterate over the transactions in the block and calculate the block txs size.
+		// This does not have access to non serializable transaction but can be used
+		// to calculate the size of the transactions handled in the block.
+		blockTxsSize := 0
+		for _, tx := range req.Txs {
+			blockTxsSize += len(tx)
+		}
+
+		telemetry.SetGaugeWithLabels(
+			[]string{blockTxsSizeMetric},
+			float32(blockTxsSize),
+			[]metrics.Label{
+				{Name: "block_height", Value: strconv.FormatInt(req.Height, 10)},
+			},
+		)
+		// Forward the request to the prepare proposal handler.
+		return prepareProposalHandler(ctx, req)
+	}
+}
+
+// initStreamingManagerWithMetrics initializes the streaming manager that listens
+// for finalize block events to capture ResponseFinalizeBlock size.
+func initStreamingManagerWithMetrics(
+	app *baseapp.BaseApp,
+	homeDir string,
+) storetypes.StreamingManager {
+	return storetypes.StreamingManager{
+		ABCIListeners: []storetypes.ABCIListener{
+			metricsABCIListener{
+				homeDir: homeDir,
+				logger:  app.Logger(),
+			},
+		},
+	}
+}
+
+// metricsABCIListener is an implementation of the StreamingManager that hooks
+// into ListenFinalizeBlock to capture ResponseFinalizeBlock size.
+type metricsABCIListener struct {
+	homeDir string
+	logger  log.Logger
+}
+
+// ListenFinalizeBlock captures the ResponseFinalizeBlock size and emits it as a
+// gauge metric.
+func (mal metricsABCIListener) ListenFinalizeBlock(
+	ctx context.Context,
+	req abci.RequestFinalizeBlock,
+	res abci.ResponseFinalizeBlock,
+) error {
+	telemetry.SetGaugeWithLabels(
+		[]string{blockResultSizeMetric},
+		float32(res.Size()),
+		[]metrics.Label{
+			{Name: "block_height", Value: strconv.FormatInt(req.Height, 10)},
+		},
+	)
+
+	var diskUsage int64
+	err := filepath.Walk(mal.homeDir, func(_ string, info os.FileInfo, err error) error {
```

Comment:
Why not use `filepath.WalkDir` instead? https://pkg.go.dev/path/filepath#WalkDir

Reply:
The more state we'll have the more this will take time to be calculated. This (at least) should not be a sync function in the block production flow.